src/serialization/container.h

```diff
@@ -103,7 +99,7 @@ bool do_serialize_container(Archive<true> &ar, C &v)
       return false;
     if (i != v.begin())
       ar.delimit_array();
-    if(!::serialization::detail::serialize_container_element(ar, const_cast<typename C::value_type&>(*i)))
+    if(!::serialization::detail::serialize_container_element(ar, (typename C::value_type&)*i))
```

Comment:
Doesn't compile, can't recall why right now.

Reply:
Ok, I can take a really good guess as to whats going on here (with help from your other response).

If the type `C` is `std::map<std::string, int>` then `C::value_type` is _always_ `std::pair<const std::string, int>` regardless of the constness of `C` itself. The STL `map` and `unordered_map` containers do not allow in-place modification of the key, only the value.

The expression `*i` is going to return `std::pair<const std::string, int>` but the `serialized_map::value_type` type is `std::pair<std::string, int>` which doesn't have this same inner-constness, thus the compilation break here with the `const_cast`. However, without the `serialized_map::value_type` "hack", then [compilation will break here](https://github.com/monero-project/monero/blob/5041de8a3b48f4db5b1a7b8f454d14bd2e3759c9/src/serialization/pair.h#L64) when trying to read from the binary archive because the `.first` element is `const`.

So this PR violates const rules, but its probably not going to invoke any bugs (right now). However, this does make the template more brittle since it removes useful compile-time checks in the C++ cast (in fact, this may have introduced a bug I haven't thought of at the moment).

Arguably that section in `pair.h` should be:

```c++
...
  typename std::remove_cv<F>::type first{}; //new
  if (!::serialization::detail::serialize_pair_element(ar, first)) // modified
    return false;
  if (!ar.stream().good())
    return false;
  ar.delimit_array();
  typename std::remove_cv<S>::type second{}; // new
  if (!::serialization::detail::serialize_pair_element(ar, second)) //modified
    return false;
  if (!ar.stream().good())
    return false;
  ar.end_array();
  p = {std::move(first), std::move(second)}; // new
}
```
then:
```c++
template<typename K, typename V>
void do_add(std::map<K, V>& c, std::pair<const K, V>&& e) { c.insert(std::move(e)); }
```
OR
```c++
template<typename K, typename V, typename T>
void do_add(std::map<K, V>& c, T&& e) { c.insert(std::forward<T>(e)); }
```

Reply:
The keys are not modified in the map though, I did check first.

Reply:
I've tried this code, but it doesn't fix, the compiler still errors out on the const_cast:

> src/serialization/container.h:103:66: error: invalid const_cast from type ‘std::pair<const crypto::public_key, cryptonote::rpc_payment::client_info>*’ to type ‘serializable_unordered_map<crypto::public_key, cryptonote::rpc_payment::client_info>::value_type*’ {aka ‘std::pair<crypto::public_key, cryptonote::rpc_payment::client_info>*’}

Could be another brick needing changing.

Reply:
> The keys are not modified in the map though, I did check first.

Yes, this should all be legal because the map itself isn't being modified. Dropping to a C-cast means the template code has less checks now and in the future though.

> I've tried this code, but it doesn't fix, the compiler still errors out on the const_cast:

I forgot to mention the part where `serialize_unordered_map` type should removed entirely for this to work - the `value_type` specified is incorrect for the `const_cast`. The quickest way to test would be to comment out the `value_type` `typedef` in `serialize_unordered_map`. If that line can be commented out, the specialized map types are not needed for serialization.

As a recap, `value_type` specified by map types have a `const` key, which caused a compilation error in the `std::pair` serialization code. The new map (`serialized_unordered_map` ...) types specified a `value_type` without a `const` key to fix that issue, but then broke the `const_cast` here. Changing the few lines in the `std::pair` serialization should mean that both of those other changes (C-cast and custom map types with a different `value_type`) should no longer be needed.